### PR TITLE
Enable, multiplication, sum and mean for vectors.

### LINKED ISSUE #1649: Cannot scale vectors with multiplication
Not sure if this in intentional, but
```Py
v = sc.scalar([1, 2, 3], unit=sc.units.m, dtype=sc.dtype.vector_3_float64)
v * (1.5 * sc.units.m)
```
throws `RuntimeError: Cannot apply operation to item dtypes vector_3_float64 float64`.

Also (probably related),
```Py
v = sc.array(dims=['x'], values=[[1, 2, 3], [4, 5, 6]], unit=sc.units.m, dtype=sc.dtype.vector_3_float64)
sc.mean(v)
```
throws `RuntimeError: Cannot apply operation to item dtypes vector_3_float64`

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
Fixes #1649

Adds overloads to allow multiplication, division, summation, mean and isfinite to be performed on variables contained 3d vectors.

### CODE REVIEW

File: variable/special_values.cpp [hunk 1]
```diff
@@ -45,4 +45,8 @@ Variable neg_inf_to_num(const VariableConstView &var,
   return transform(var, replacement, element::negative_inf_to_num);
 }
 
+Variable isfinite(const VariableConstView &var) {
+  return transform(var, element::isfinite);
+}
```
Comment (SimonHeybrock) [inline variable/special_values.cpp:50]: Why not generated any more?

File: variable/include/scipp/variable/transform.h [hunk 2]
```diff
@@ -349,6 +349,15 @@ static void do_transform(Op op, Out &&out, Tuple &&processed, const Arg &arg,
...
                                    " not supported.");
+    } else if constexpr (
+        std::is_base_of_v<
+            core::transform_flags::
+                expect_no_in_variance_if_out_cannot_have_variance_t,
+            Op> &&
+        !core::canHaveVariances<typename Out::value_type>()) {
+      throw except::VariancesError(
+          "Variances in argument " + std::to_string(std::tuple_size_v<Tuple>) +
+          " not supported. As output dtype cannot have variances");
```
Comment (SimonHeybrock) [inline variable/include/scipp/variable/transform.h:360]:
```suggestion
          " not supported as output dtype cannot have variances");
```

File: variable/test/sum_test.cpp [hunk 3]
```diff
@@ -58,3 +58,23 @@ TEST_F(SumTest, sum_in_place_bool_incorrect_out_type) {
   auto out = makeVariable<float>(Dims{Dim::Y}, Shape{2});
   EXPECT_THROW(sum(var_bool, Dim::X, out), except::TypeError);
 }
+
+TEST(VectorReducTest, sum_vector) {
```
Comment (SimonHeybrock) [inline variable/test/sum_test.cpp:62]:
```suggestion
TEST(VectorReduceTest, sum_vector) {
```

File: variable/test/sum_test.cpp [hunk 4]
```diff
@@ -58,3 +58,23 @@ TEST_F(SumTest, sum_in_place_bool_incorrect_out_type) {
...
+  const auto vector_var = makeVariable<Eigen::Vector3d>(
+      Dims{Dim::X}, Shape{2}, units::m,
+      Values{Eigen::Vector3d{1, 2, 3}, Eigen::Vector3d{4, 5, 6}});
+  const auto expected = makeVariable<Eigen::Vector3d>(
+      Dims{}, Shape{1}, units::m, Values{Eigen::Vector3d{5, 7, 9}});
+  auto summed = sum(vector_var, Dim::X);
+  EXPECT_EQ(vector_var, expected);
+}
+
+TEST(VectorReducTest, mean_vector) {
```
Comment (SimonHeybrock) [inline variable/test/sum_test.cpp:72]:
```suggestion
TEST(VectorReduceTest, mean_vector) {
```

File: core/include/scipp/core/element/arithmetic.h [hunk 5]
```diff
@@ -86,7 +94,12 @@ constexpr auto plus =
 constexpr auto minus =
     overloaded{add_types_t{}, [](const auto a, const auto b) { return a - b; }};
 constexpr auto times = overloaded{
-    times_types_t{}, [](const auto a, const auto b) { return a * b; }};
+    times_types_t{},
+    transform_flags::expect_no_in_variance_if_out_cannot_have_variance,
+    [](const auto a, const auto b) {
+      return a * b;
+    } // namespace scipp::core::element
+};
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element/arithmetic.h:102]: Are the in-place versions supported already, or will this be added in the future?

File: core/include/scipp/core/element/arithmetic.h [hunk 6]
```diff
@@ -70,10 +70,18 @@ struct add_types_t {
...
+                     std::tuple<std::tuple<float, Eigen::Vector3d>>(),
+                     std::tuple<std::tuple<int64_t, Eigen::Vector3d>>(),
+                     std::tuple<std::tuple<int32_t, Eigen::Vector3d>>(),
+                     std::tuple<std::tuple<Eigen::Vector3d, double>>(),
+                     std::tuple<std::tuple<Eigen::Vector3d, float>>(),
+                     std::tuple<std::tuple<Eigen::Vector3d, int64_t>>(),
+                     std::tuple<std::tuple<Eigen::Vector3d, int32_t>>()));
 };
 
 struct divide_types_t {
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element/arithmetic.h:87]: I suppose `vector / scalar` should also be supported?

File: variable/include/scipp/variable/transform.h [hunk 7]
```diff
@@ -349,6 +349,15 @@ static void do_transform(Op op, Out &&out, Tuple &&processed, const Arg &arg,
       throw except::VariancesError("Variances in argument " +
                                    std::to_string(std::tuple_size_v<Tuple>) +
                                    " not supported.");
+    } else if constexpr (
+        std::is_base_of_v<
+            core::transform_flags::
+                expect_no_in_variance_if_out_cannot_have_variance_t,
+            Op> &&
+        !core::canHaveVariances<typename Out::value_type>()) {
```
Comment (SimonHeybrock) [inline variable/include/scipp/variable/transform.h:357]: Can you add a test for this in `transform_test.cpp`? Other flags are tested there as well.

**Review by SimonHeybrock — COMMENTED** [review]
Looks mostly goods. Don't forget to mention the issue in the PR so it will get closed, if there is one.

File: variable/special_values.cpp (hunk 1 above)
Comment (Matthew-Andrew) [reply variable/special_values.cpp:50]: make_unary{isfinite} was causing compile errors as it attempted to set the bool return value to a vector. I didn't think the inplace version of this operator was needed so this seemed like the simplest solution. If the inplace version is needed I will need to take a step back and figure out how to get it to work.

File: core/include/scipp/core/element/arithmetic.h (hunk 5 above)
Comment (Matthew-Andrew) [reply core/include/scipp/core/element/arithmetic.h:102]: In place operations are already supported.

File: variable/special_values.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply variable/special_values.cpp:50]: The codegen uses the `NO_OUT` arg, so the inplace version should not even be generated? 

I don't think we should just silently remove functions, so yes, we need to get to the bottom of this, understanding why this failed, despite apparently being disabled by `NO_OUT`?

Comment (Matthew-Andrew) [reply variable/special_values.cpp:50]: I think NO_OUT only affects the python function generation and not the C++ side. Adding something like the below line from python_unary.cpp to variable_unary.cpp may be the solution.

`if constexpr(std::is_same_v<T, Variable> && @GENERATE_OUT@)`

Comment (SimonHeybrock) [reply variable/special_values.cpp:50]: Right. Can you try what you suggest?

File: core/include/scipp/core/element/arithmetic.h [hunk 8]
```diff
@@ -76,32 +76,52 @@ struct add_types_t {
...
-                               std::is_integral_v<decltype(b)>)
-                   return static_cast<double>(a) / b;
-                 else
-                   return a / b;
-               }};
+    times_types_t{},
+    transform_flags::expect_no_in_variance_if_out_cannot_have_variance,
+    [](const auto a, const auto b) {
+      return a * b;
+    } // namespace scipp::core::element
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element/arithmetic.h:112]: Are you an old clang-format version? We used to have this formatting issue, but is disappeared a while ago.

File: templates/variable_unary.cpp.in [hunk 9]
```diff
@@ -15,8 +15,13 @@ Variable @NAME@(const VariableConstView &var) {
...
 
 VariableView @NAME@(const VariableConstView &var, const VariableView &out) {
-  transform_in_place(out, var, assign_unary{element::@OPNAME@});
+  if constexpr (@GENERATE_OUT@) {
+  transform_in_place(
+      out, var, assign_unary { element::@OPNAME@ });
   return out;
+  } else {
+    throw std::logic_error("Inplace operation not supported for @NAME@");
+  }
```
Comment (SimonHeybrock) [inline templates/variable_unary.cpp.in:24]: I would suggest to simply not generate this function in this case, rather than throwing.

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Looks good. Could you add something to the release notes about this?

File: variable/test/operations_test.cpp [hunk 10]
```diff
@@ -184,9 +184,9 @@ TEST(Variable, operator_plus_unit_fail) {
   auto b = makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{1.0, 2.0},
                                 Variances{3.0, 4.0});
   b.setUnit(units::s);
-  ASSERT_ANY_THROW(a + b);
+  ASSERT_ANY_THROW([[maybe_unused]] auto c = a + b);
   b.setUnit(units::m);
-  ASSERT_NO_THROW(a + b);
+  ASSERT_NO_THROW([[maybe_unused]] auto c = a + b);
```
Comment (SimonHeybrock) [inline variable/test/operations_test.cpp:189]: Note that we have these helpers for this purpose: https://github.com/scipp/scipp/blob/92266385399dec2b40ad6bccf1a726db766619b6/test/test_macros.h#L55-L61

**Review by SimonHeybrock — APPROVED** [review]

File: variable/test/operations_test.cpp [hunk 11]
```diff
@@ -184,9 +184,9 @@ TEST(Variable, operator_plus_unit_fail) {
   auto b = makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{1.0, 2.0},
                                 Variances{3.0, 4.0});
   b.setUnit(units::s);
-  ASSERT_ANY_THROW([[maybe_unused]] auto c = a + b);
+  ASSERT_ANY_THROW_DISCARD(a + b);
   b.setUnit(units::m);
-  ASSERT_NO_THROW([[maybe_unused]] auto c = a + b);
+  ASSERT_ANY_THROW_DISCARD(a + b);
```
Comment (SimonHeybrock) [inline variable/test/operations_test.cpp]: - Why not check for `UnitError`?
- Was the change from `NO_THROW` intentional?

File: core/include/scipp/core/element/creation.h [hunk 12]
```diff
@@ -24,6 +24,8 @@ constexpr auto zeros_not_bool_like =
                  using T = std::decay_t<decltype(x)>;
                  if constexpr (std::is_same_v<T, bool>)
                    return int64_t{0};
+                 else if constexpr (std::is_same_v<Eigen::Vector3d, T>)
```
Comment (OwenArnold) [inline core/include/scipp/core/element/creation.h:27]: Might have been safer to try to detect an Eigen templated base class and apply T::zero on that.

pseudo 
```
template<typename ...T>
struct is_eigen_matrix : std::false_type {
};

template <typename ...T>
struct is_eigen_matrix<Eigen::Matrix<T...>> : std::true_type {
};
```

Comment (OwenArnold) [reply core/include/scipp/core/element/creation.h:27]: @Matthew-Andrew in fact, I think exactly this is being done in `data_model.h`

**Review by OwenArnold — APPROVED** [review]

**Comment by OwenArnold** [comment]
This is working now. I'll open up another PR with some cleanup and documentation, release notes.

**Pull request merged**